### ISSUE TITLE
Make the update function return the model.

### ISSUE THREAD

**Issue by jeandudey** [antoyo/relm#15]
I was reading the README and saw this:

>Your program might be slower when using this attribute because the code generation is simple. For instance, the following code
>```rust
>fn update(&mut self, event: Msg, model: &mut Model) {
>    for _ in 0..100 {
>        model.counter += 1;
>    }
>}
>```
>
> will generate this function:
>
> ```rust
>fn update(&mut self, event: Msg, model: &mut Model) {
>    for _ in 0..100 {
>        model.counter += 1;
>        self.label1.set_text(&model.counter.to_string());
>    }
>}
>```

I think that it is too magical and users of the library aren't going to pay attention to what the readme says (like me) and make that mistake.

And I wondered if instead of borrowing the model (`&mut Model`), take ownership of it and let the user return a modified one. e.g:

```rust
fn update(&mut self, event: Msg, model: Model) -> Model {
    for _ in 0..100 {
        model.counter += 1;
    }
    model
}
```

Another way to solve it would be creating a `do_update!` or `make_visible!` macro, that modifies the widget with the updated model, e.g:

```rust
fn update(&mut self, event: Msg, model: &mut Model) {
    for _ in 0..100 {
        model.counter += 1;
    }
    make_visible!(model)
}
```

Also, nice project!

**Comment by antoyo** [comment]
Your first suggestion was the way used in the very first versions of `relm` and I switch to the more idiomatic `&mut Model` and it will probably come back to be an attribute to `self`.

In any case, neither of these solutions solve the issue with efficiently updating the view.

Are you worried about performance?

One solution that I thought to produce a more efficient version is currently described in [a comment](https://github.com/antoyo/relm/blob/2e003f609e6e0f6e56548e090821c190d6a26ef2/relm-attributes/src/lib.rs#L294-L295).

This idea can be schematized as follow:
```
            *
          /   \
        A       A
          \    /
            1
```
where `1` updates the properties bound to the model attribute `A`.
This is the case of a condition: this will result in generating one call to `set_property()` instead of 2.
I believe the same logic can be applied to loops so that the view is only updated after the loop.

However, it is complex to implement, so it will probably take some time before I do it.

**Comment by jeandudey** [comment]
>Are you worried about performance?
No, i'm not worried about performance, i just thought that it was to magical and slightly hard to know was going under the hood.

**Comment by antoyo** [comment]
If you're not worried about performance, then I believe you want to view to be updated as a whole with the new model, right?
I think recreating the view each time is a huge cost, so it will degrade the performance in a visible way when creating complex view.

One solution (that is currently impossible because not implemented in GTK+ yet) is to switch to GTK Scene Graph. As the API is not yet available, I'm not sure how it will look, but that will fix this issue, I think, amongs others.
